pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+//! # Lottery Pallet
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{BuildGenesisConfig, Randomness},
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, Saturating, Zero},
+	DispatchResult, Percent,
+};
+use sp_std::{fmt::Debug, marker::PhantomData, prelude::*, vec::Vec};
+
+use primitives::Role;
+use traits::{BasicAccounting, GetTreasury, ManageRoles};
+
+mod mock;
+mod tests;
+
+pub mod weights;
+pub use weights::*;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub use module::*;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+	use sp_runtime::traits::BlockNumberProvider;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+
+		/// The balance type
+		type Balance: Member
+			+ Parameter
+			+ MaxEncodedLen
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ From<u128>
+			+ sp_std::iter::Sum;
+
+		type RoleManager: ManageRoles<Self::AccountId>;
+
+		type Bank: BasicAccounting<Self::AccountId, Self::Balance> + GetTreasury<Self::AccountId>;
+
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
+
+		type Randomness: Randomness<Self::Hash, BlockNumberFor<Self>>;
+
+		#[pallet::constant]
+		type LotteryPayoutPeriod: Get<BlockNumberFor<Self>>;
+
+		#[pallet::constant]
+		type PrizePoolAccount: Get<Self::AccountId>;
+
+		#[pallet::constant]
+		type TaxRate: Get<Percent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The prize split total is not equal to one.
+		InvalidPrizeSplitTotal,
+		/// The account role does not equal to the expected role.
+		IncorrectRole,
+		/// The ticket price is not set.
+		TicketPriceNotSet,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Set the prize split.
+		PrizeSplitUpdated { split: Vec<Percent> },
+		/// Set the ticket price.
+		TicketPriceUpdated { old: T::Balance, new: T::Balance },
+		/// Customer bought numbers of tickets.
+		TicketsBought { id: T::AccountId, number: u32 },
+		/// Customer won the lottery fund and paid tax.
+		LotteryWon { user: T::AccountId, won_fund: T::Balance, tax: T::Balance },
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn prize_split)]
+	pub type PrizeSplit<T: Config> = StorageValue<_, Vec<Percent>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn ticket_price)]
+	pub type TicketPrice<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn tickets)]
+	pub type PlayersAndLotteries<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, u32>;
+
+	#[pallet::storage]
+	pub(crate) type RandomSeed<T: Config> = StorageValue<_, u32, ValueQuery>;
```

Comment:
I think we can use the current Block number as the random seed (as it is non-repeating).
This saves us from this extra storage.